mlir/lib/Dialect/SCF/Transforms/ForallToFor.cpp

```diff
@@ -0,0 +1,92 @@
+//===- ForallToFor.cpp - scf.forall to scf.for loop conversion ------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Transforms SCF.ForallOp's into SCF.ForOp's.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/SCF/Transforms/Passes.h"
+
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+#define GEN_PASS_DEF_SCFFORALLTOFORLOOP
+#include "mlir/Dialect/SCF/Transforms/Passes.h.inc"
+} // namespace mlir
+
+using namespace llvm;
+using namespace mlir;
+using scf::ForallOp;
+using scf::ForOp;
+
+LogicalResult
+mlir::scf::forallToForLoop(RewriterBase &rewriter, scf::ForallOp forallOp,
+                           SmallVector<Operation *> *results = nullptr) {
+  rewriter.setInsertionPoint(forallOp);
+
+  if (!forallOp.getOutputs().empty()) {
+    return forallOp.emitOpError()
+           << "unsupported shared outputs (didn't bufferize?)";
+  }
+
+  SmallVector<OpFoldResult> lbs = forallOp.getMixedLowerBound();
+  SmallVector<OpFoldResult> ubs = forallOp.getMixedUpperBound();
+  SmallVector<OpFoldResult> steps = forallOp.getMixedStep();
+
+  auto loc = forallOp.getLoc();
+  SmallVector<Value> ivs;
+  for (auto &&[lb, ub, step] : llvm::zip(lbs, ubs, steps)) {
+    Value lbValue = getValueOrCreateConstantIndexOp(rewriter, loc, lb);
+    Value ubValue = getValueOrCreateConstantIndexOp(rewriter, loc, ub);
+    Value stepValue = getValueOrCreateConstantIndexOp(rewriter, loc, step);
+    auto loop =
+        rewriter.create<ForOp>(loc, lbValue, ubValue, stepValue, ValueRange(),
+                               [](OpBuilder &, Location, Value, ValueRange) {});
+    if (results)
```

Comment:
You can just do `mergeBlocks` (like here : https://github.com/shark-infra/llvm-project/blob/08163cd9d82690e808c28515523b5fd0923d7b38/mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp#L345)

Reply:
How exactly should I be using `mergeBlocks`? The forall block only needs to be inlined/merged with the the innermost for loop so I am not seeing how to use `mergeBlocks` within this loop?

Reply:
Ok, maybe you can just used `buildLoopNest` and then just do `mergeBlocks` on the inner most (i.e. just build an empty inner-most loop and then merge the body of the nested scf.for with the body of the scf.forall.

Reply:
I updated to using `buildLoopNest` but this inserts a yield terminator into every loop's body already so the innermost loop body is not empty and `mergeBlocks` will error as this results in the yield op not being the last operation in the block. I could probably get around this by merging and moving, but not sure whether that's better than using `inlineBlockBefore`?